File: Depth-first_search(DFS)/python/02_유기농배추.py
import sys
sys.setrecursionlimit(10000)

T = int(input())
mx = [1, -1, 0, 0]
my = [0, 0, -1, 1]

def dfs(x, y):
    visited[x][y] = True
    for i in range(4):
        nx = x + mx[i]
        ny = y + my[i]
        if nx < 0 or nx >= n or ny < 0 or ny >= m:
            continue
        if adj[nx][ny] == 1 and not(visited[nx][ny]):
            dfs(nx, ny)

result = [0] * T

for t in range(T):
    m, n, k = map(int, input().split())
    adj = [[0] * (m) for _ in range(n)]
    visited = [[False] * (m) for _ in range(n)]
    for _ in range(k):
        x, y = map(int, input().split())
        adj[y][x] = 1

    for i in range(n):
        for j in range(m):
            if adj[i][j] == 1 and not(visited[i][j]):
                dfs(i, j)
                result[t] += 1

for i in range(T):
    print(result[i])

File: Depth-first_search(DFS)/c++/02_유기농배추.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/1012
// 1

#include <iostream>
#include <cstring>

using namespace std;

int matrix[51][51];
bool visited[51][51];
int M, N, K;
int mx[4] = {-1, 1, 0, 0};
int my[4] = {0, 0, -1, 1};

void dfs(int x, int y)
{
    visited[x][y] = true;

    for (int i = 0; i < 4; i++) {
        int nx = x + mx[i];
        int ny = y + my[i];

        if (nx < 0 || nx >= N || ny < 0 || ny >= M)
            continue;
        if (!visited[nx][ny] && matrix[nx][ny])
            dfs(nx, ny);
    }
}

int main()
{
    int test_case;
    cin >> test_case;
    for (int t = 0; t < test_case; t++) {
        memset(matrix, 0, sizeof(matrix));
        memset(visited, false, sizeof(visited));
        cin >> M >> N >> K;

        for (int i = 0; i < K; i++) {
            int x, y;
            cin >> y >> x;
            matrix[x][y] = 1;
        }

        int result = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (!visited[i][j] && matrix[i][j]) {
                    result++;
                    dfs(i, j);
                }
            }
        }
        cout << result << '\n';
    }
}